lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -1872,26 +1870,35 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
           }
         }
 
-        // Handle cases that axes are explicitly specified.
-        // Extract the axes values from the axes operand.
-        // This really shouldn't happen but it helps pass weird tests.
-        // TODO: Derive the chosen axes from the data type and final result type
-        // instead of using the dynamic axes at operand[1].
-        if (!binder.tensorOperandAtIndex(axes, 1)) {
-          Torch::BaseTensorType axesType =
-              axes.getType().cast<Torch::BaseTensorType>();
-          auto sizes = axesType.getSizes();
-          for (int i = 0; i < sizes[0]; i++) {
-            Value selectIndex = rewriter.create<Torch::ConstantIntOp>(
-                binder.getLoc(), rewriter.getI64IntegerAttr(i));
-            Value extract = rewriter.create<Torch::AtenSelectIntOp>(
-                binder.getLoc(),
-                axesType.getWithSizesAndDtype(llvm::SmallVector<int64_t>{1},
-                                              axesType.getOptionalDtype()),
-                axes, zero, selectIndex);
-            Value dim = rewriter.create<Torch::AtenItemOp>(binder.getLoc(),
-                                                           torchIntTy, extract);
-            axesList.push_back(dim);
+        // Derive the chosen axes from the data type and final result type
+        // Fix --convert-torch-to-linalg bug:  `dim` argument must be a
+        // constant int list or None
+        auto resultTypeSizes = resultType.getSizes();
+        int resultRank = resultTypeSizes.size();
+        if (!keepDims) {
+          // Handle the keepDimsBool == False case:
+          // [2,3,4] -> [2,4] when axes=1
+          int j = 0;
+          for (int i = 0; i < rank; i++) {
+            if (resultRank && j < resultRank &&
+                dataSize[i] == resultTypeSizes[j]) {
```

Comment:
How could you identify which dimension the reduction is done over for a `[4, 4, 4] -> [4]`? You can't without looking at the axis. You will need to use that information somehow.

Reply:
No. I cannot identify this cases. Let me think about it.

Reply:
Then this implementation is not able to fix all the cases. Either we add a special check for this kind of case. 
Or we back to the previous implementation. But the previous implementation, after we select the dynamic axis, it will fail when lowering to linalg `dim argument must be a constant int list or None`. I feel kind of stuck what to do next.

Reply:
@rsuderman Could you provide some guidance on it?

Reply:
I honestly wouldn't worry about cases where we can't fold the dim value.  Its a bit nonsensical if we don't know what the value is. We could do a shape determined value, i.e. infer from result types when the value is not foldable, but I don't think there is much value in the added complexity.

Reply:
` Its a bit nonsensical if we don't know what the value is.` Do you mean the operand1/dimtensor is not useless? This is what I am trying to do in this patch to derive the shape from the result types. But in the first comment, you said `We should not assume the values do not matter.`

Reply:
Just added check for input shape dims are distinct when dynamic axes given